src/ripple/core/sql_backend/SQLInterface_sqlite.cpp

```diff
@@ -0,0 +1,1365 @@
+    firstLedgerSeq(LedgerIndex shardIndex)
+    {
+        return app_.getShardStore()->firstLedgerSeq(shardIndex);
+    }
+
+    LedgerIndex
+    lastLedgerSeq(LedgerIndex shardIndex)
+    {
+        return app_.getShardStore()->lastLedgerSeq(shardIndex);
+    }
+
+    bool
+    existsLedger()
+    {
+        return !!lgrdb_;
+    }
+
+    bool
+    existsTransaction()
+    {
+        return !!txdb_;
+    }
+
+    auto
+    checkoutLedger()
+    {
+        return lgrdb_->checkoutDb();
+    }
+
+    auto
+    checkoutTransaction()
+    {
+        return txdb_->checkoutDb();
+    }
+
+    bool
+    doLedger(
+        LedgerIndex ledgerSeq,
+        std::function<bool(soci::session& session, LedgerIndex index)> const&
+            callback)
+    {
+        return app_.getShardStore()->callForLedgerSQL(ledgerSeq, callback);
+    }
+
+    bool
+    doTransaction(
+        LedgerIndex ledgerSeq,
+        std::function<bool(soci::session& session, LedgerIndex index)> const&
+            callback)
+    {
+        return app_.getShardStore()->callForTransactionSQL(ledgerSeq, callback);
+    }
+
+    bool
+    iterateLedgerForward(
+        LedgerIndex firstIndex,
+        std::function<bool(soci::session& session, LedgerIndex index)> const&
+            callback)
+    {
+        return app_.getShardStore()->iterateLedgerSQLsForward(firstIndex, callback);
+    }
+
+    bool
+    iterateTransactionForward(
+        LedgerIndex firstIndex,
+        std::function<bool(soci::session& session, LedgerIndex index)> const&
+            callback)
+    {
+        return app_.getShardStore()->iterateLedgerSQLsForward(firstIndex, callback);
+    }
+
+    bool
+    iterateLedgerBack(
+        LedgerIndex firstIndex,
+        std::function<bool(soci::session& session, LedgerIndex index)> const&
+            callback)
+    {
+        return app_.getShardStore()->iterateLedgerSQLsBack(firstIndex, callback);
+    }
+
+    bool
+    iterateTransactionBack(
+        LedgerIndex firstIndex,
+        std::function<bool(soci::session& session, LedgerIndex index)> const&
+            callback)
+    {
+        return app_.getShardStore()->iterateLedgerSQLsBack(firstIndex, callback);
+    }
+};
+
+bool
+SQLInterface_sqlite::makeLedgerDBs(
+    Config const& config,
+    DatabaseCon::Setup const& setup,
+    DatabaseCon::CheckpointerSetup const& checkpointerSetup)
+{
+    auto [res, tx, lgr] = ripple::makeLedgerDBs(
+        config,
+        setup,
+        checkpointerSetup);
+    txdb_ = std::move(tx);
+    lgrdb_ = std::move(lgr);
+    return res;
+}
+
+std::optional<LedgerIndex>
+SQLInterface_sqlite::getMinLedgerSeq()
+{
+    /* if databases exists, use it */
+    if (existsLedger())
+    {
+        auto db = checkoutLedger();
+        return ripple::getMinLedgerSeq(*db, TableType::Ledgers);
+    }
+
+    /* else use shard databases */
+    std::optional<LedgerIndex> res;
+    iterateLedgerForward(
+        (LedgerIndex)-1, [&](soci::session& session, LedgerIndex index) {
```

Comment:
I wouldn't use `-1` as a magic value here - especially since `LedgerIndex` is an unsigned type. If we can't use `0` because the implementation uses `>` instead of `>=` (although it's not clear to me why we use `>`), then can we pass in an `optional` instead?

Reply:
Done.